src/domain/opendatafabric/src/identity/grammar.rs

```diff
@@ -114,6 +153,60 @@ impl Grammar {
         Some((&s[0..h.len() + hh.len()], tt))
     }
 
+    // StartElement = HostName ("." | "-")?
+    pub fn match_hostname_pattern_start_element(s: &str) -> Option<(&str, &str)> {
+        let (h1, t1) = Self::match_hostname(s)?;
+        let (h2, t2) = Self::match_opt(t1, |s| Self::match_one_of_chars(s, &['.', '-']))?;
+        Some((&s[..h1.len() + h2.len()], t2))
+    }
+
+    // MiddleElement = (
+    //     (("." | "-")? HostName ("." | "-")?)
+    //   | ("." | "-")
+    // ) non-EOF
+    pub fn match_hostname_pattern_middle_element(s: &str) -> Option<(&str, &str)> {
+        let (h1, t1) = Self::match_alt(
+            s,
+            |s| {
+                let (h1, t1) = Self::match_opt(s, |s| Self::match_one_of_chars(s, &['.', '-']))?;
+                let (h2, t2) = Self::match_hostname(t1)?;
+                let (h3, t3) = Self::match_opt(t2, |s| Self::match_one_of_chars(s, &['.', '-']))?;
+                Some((&s[..h1.len() + h2.len() + h3.len()], t3))
+            },
+            |s| Self::match_one_of_chars(s, &['.', '-']),
+        )?;
+        let (h2, t2) = Self::match_non_eof(t1)?;
+        Some((&s[..h1.len() + h2.len()], t2))
+    }
+
+    // EndElement = ("." | "-")? HostName
+    pub fn match_hostname_pattern_end_element(s: &str) -> Option<(&str, &str)> {
+        let (h1, t1) = Self::match_opt(s, |s| Self::match_one_of_chars(s, &['.', '-']))?;
+        let (h2, t2) = Self::match_hostname(t1)?;
+        Some((&s[..h1.len() + h2.len()], t2))
+    }
+
+    // Pattern = StartElement? ("%"+ MiddleElement)* "%"+ EndElement?
+    pub fn match_hostname_pattern(s: &str) -> Option<(&str, &str)> {
```

Comment:
I'd like to revisit this again when we have a much better test coverage.

But my gut reaction is that this might be an overkill:
- Invalid pattern will never result in us creating some incorrectly named object in the system - thanks to existing grammar for `DatasetName` we know for sure that all dataset names are already valid `Hostname`s
- The worst think that can happen in case of an invalid patterns like `%a..b` is that **it will simply not match anything**
- This makes me question the point of strict pattern validation. If user has invalid elements in the pattern like `a..b` is it any better to complain about invalid pattern than to return empty result?

CC @zaychenko-sergei

Reply:
Discussed this with @zaychenko-sergei further:

Having this grammar makes behavior more consistent with `DatasetRef` where invalid references fail validation rather than not match anything.

My concern was primarily the complexity of expressing this grammar, but given that:
- it's already been written
- implementation itself is not that complex
- and can be easily extended to other aspects of `LIKE` syntax (e.g. matching any character by underscore `_`)

Then there is no reason not to keep it.